In [44]:
import pandas as pd
import numpy as np

In [45]:
df_flights = pd.read_csv("/Users/rezanaiman/Desktop/Fall 2024/Stat captson/ICR project/final_merged_2024.csv", low_memory = False)

In [46]:
df_flights.describe()

YEAR   QUARTER     MONTH   DAY_OF_MONTH    DAY_OF_WEEK  \
count  121884.0  121884.0  121884.0  121884.000000  121884.000000   
mean     2024.0       1.0       1.0      16.261618       3.813651   
std         0.0       0.0       0.0       8.803848       2.014112   
min      2024.0       1.0       1.0       1.000000       1.000000   
25%      2024.0       1.0       1.0       9.000000       2.000000   
50%      2024.0       1.0       1.0      16.000000       4.000000   
75%      2024.0       1.0       1.0      24.000000       6.000000   
max      2024.0       1.0       1.0      31.000000       7.000000   

       OP_CARRIER_AIRLINE_ID  ORIGIN_AIRPORT_ID  ORIGIN_AIRPORT_SEQ_ID  \
count          121884.000000      121884.000000           5.800000e+04   
mean            20004.282080       12329.953062           1.200599e+06   
std               382.601389        1312.190489           1.316687e+05   
min             19393.000000       10135.000000           1.013506e+06   
25%             19790.000000       11292.000000           1.129202e+06   
50%             19977.000000       12339.000000           1.129202e+06   
75%             20363.000000       12953.000000           1.288904e+06   
max             20452.000000       16869.000000           1.686902e+06   

       ORIGIN_CITY_MARKET_ID  DEST_AIRPORT_ID  ...      ARR_TIME  \
count          121884.000000    121884.000000  ...  55135.000000   
mean            31498.463539     12393.453004  ...   1506.383477   
std              1257.734873      1352.094184  ...    493.493147   
min             30135.000000     10135.000000  ...      1.000000   
25%             30325.000000     11292.000000  ...   1149.000000   
50%             31454.000000     12391.000000  ...   1529.000000   
75%             31703.000000     13204.000000  ...   1910.000000   
max             35582.000000     16869.000000  ...   2400.000000   

           ARR_DELAY  ARR_DELAY_NEW      CANCELLED      AIR_TIME  FLIGHTS  \
count  115594.000000  115594.000000  121884.000000  55021.000000  58000.0   
mean       10.678331      20.841748       0.048579    117.496938      1.0   
std        69.360097      65.135290       0.214987     54.335044      0.0   
min       -87.000000       0.000000       0.000000     11.000000      1.0   
25%       -18.000000       0.000000       0.000000     84.000000      1.0   
50%        -6.000000       0.000000       0.000000    112.000000      1.0   
75%        15.000000      15.000000       0.000000    140.000000      1.0   
max      1816.000000    1816.000000       1.000000    476.000000      1.0   

            DISTANCE  WEATHER_DELAY          tmpf          alti  
count  121884.000000   29224.000000  71585.000000  71585.000000  
mean      875.633069       7.105051     30.372436     29.972379  
std       547.996762      49.513034     13.581435      0.286750  
min        73.000000       0.000000    -16.000000     28.920000  
25%       524.000000       0.000000     24.000000     29.820000  
50%       776.000000       0.000000     32.000000     29.990000  
75%      1074.000000       0.000000     39.000000     30.160000  
max      4983.000000    1804.000000     63.000000     30.630000  

[8 rows x 27 columns]

In [47]:
df_flights.shape

(121884, 69)

In [48]:
df_flights.head(15)

YEAR  QUARTER  MONTH  DAY_OF_MONTH  DAY_OF_WEEK        FL_DATE  \
0   2024        1      1             1            1  1/1/2024 0:00   
1   2024        1      1             1            1  1/1/2024 0:00   
2   2024        1      1             1            1  1/1/2024 0:00   
3   2024        1      1             1            1  1/1/2024 0:00   
4   2024        1      1             1            1  1/1/2024 0:00   
5   2024        1      1             1            1  1/1/2024 0:00   
6   2024        1      1             1            1  1/1/2024 0:00   
7   2024        1      1             1            1  1/1/2024 0:00   
8   2024        1      1             1            1  1/1/2024 0:00   
9   2024        1      1             1            1  1/1/2024 0:00   
10  2024        1      1             1            1  1/1/2024 0:00   
11  2024        1      1             1            1  1/1/2024 0:00   
12  2024        1      1             1            1  1/1/2024 0:00   
13  2024        1      1             1            1  1/1/2024 0:00   
14  2024        1      1             1            1  1/1/2024 0:00   

   OP_UNIQUE_CARRIER  OP_CARRIER_AIRLINE_ID TAIL_NUM  ORIGIN_AIRPORT_ID  ...  \
0                 AA                  19805   N102UW              11292  ...   
1                 AA                  19805   N102UW              14107  ...   
2                 AA                  19805   N141NN              11298  ...   
3                 AA                  19805   N144AN              11292  ...   
4                 AA                  19805   N144AN              11298  ...   
5                 AA                  19805   N149AN              11292  ...   
6                 AA                  19805   N150UW              11057  ...   
7                 AA                  19805   N150UW              11292  ...   
8                 AA                  19805   N155NN              11292  ...   
9                 AA                  19805   N160AN              11298  ...   
10                AA                  19805   N3014R              11298  ...   
11                AA                  19805   N3014R              11503  ...   
12                AA                  19805   N338RS              13303  ...   
13                AA                  19805   N341SV              11292  ...   
14                AA                  19805   N355PU              13930  ...   

    skyl3  skyl4 wxcodes peak_wind_gust peak_wind_drct peak_wind_time   feel  \
0       M      M       M              M              M              M  12.28   
1     NaN    NaN     NaN            NaN            NaN            NaN    NaN   
2     NaN    NaN     NaN            NaN            NaN            NaN    NaN   
3       M      M       M              M              M              M     20   
4     NaN    NaN     NaN            NaN            NaN            NaN    NaN   
5       M      M       M              M              M              M  16.72   
6     NaN    NaN     NaN            NaN            NaN            NaN    NaN   
7   22000      M       M              M              M              M  38.19   
8       M      M       M              M              M              M  13.33   
9     NaN    NaN     NaN            NaN            NaN            NaN    NaN   
10    NaN    NaN     NaN            NaN            NaN            NaN    NaN   
11      M      M       M              M              M              M     20   
12    NaN    NaN     NaN            NaN            NaN            NaN    NaN   
13    NaN    NaN     NaN            NaN            NaN            NaN    NaN   
14    NaN    NaN     NaN            NaN            NaN            NaN    NaN   

                                                metar  snowdepth  \
0   KDEN 011153Z 21006KT 10SM FEW200 M06/M10 A3011...          M   
1                                                 NaN        NaN   
2                                                 NaN        NaN   
3   KDEN 011253Z 00000KT 10SM FEW220 M07/M11 A301

In [49]:
colorado_count = df_flights[df_flights["ORIGIN_AIRPORT_ID"].isin([10325, 11292, 11784])].shape[0]


In [50]:
print("Number of observation in the dataset of Colorado Airports ", colorado_count)
diff  = 121884 - 28301
diff

Number of observation in the dataset of Colorado Airports  28301


93583

 
- 10325 is the Airport ID for Denver Centenial Airport
- 11292 is the Airport ID for Denver International Airport
- 11784 is the Colorado Air and Spcae Port

The combination of these airports makes up the 28301 observations in Colorado. So 93583 of the observations belong to New York.

In [43]:
df_flights.shape
df_flights_clean = df_flights.dropna(subset=df_flights.columns[27:67],how = 'any')
df_flights_clean

Empty DataFrame
Columns: [YEAR, QUARTER, MONTH, DAY_OF_MONTH, DAY_OF_WEEK, FL_DATE, OP_UNIQUE_CARRIER, OP_CARRIER_AIRLINE_ID, TAIL_NUM, ORIGIN_AIRPORT_ID, ORIGIN_AIRPORT_SEQ_ID, ORIGIN_CITY_MARKET_ID, ORIGIN, ORIGIN_CITY_NAME, ORIGIN_STATE_ABR, ORIGIN_STATE_NM, DEST_AIRPORT_ID, DEST_AIRPORT_SEQ_ID, DEST_CITY_MARKET_ID, DEST, DEST_CITY_NAME, DEST_STATE_ABR, DEST_STATE_NM, CRS_DEP_TIME, DEP_TIME, DEP_DELAY, DEP_DELAY_NEW, DEP_TIME_BLK, CRS_ARR_TIME, ARR_TIME, ARR_DELAY, ARR_DELAY_NEW, ARR_TIME_BLK, CANCELLED, CANCELLATION_CODE, AIR_TIME, FLIGHTS, DISTANCE, WEATHER_DELAY, departure_datetime, rounded_departure, station, valid, tmpf, dwpf, relh, drct, sknt, p01i, alti, mslp, vsby, gust, skyc1, skyc2, skyc3, skyc4, skyl1, skyl2, skyl3, skyl4, wxcodes, peak_wind_gust, peak_wind_drct, peak_wind_time, feel, metar, snowdepth, rounded_valid]
Index: []

[0 rows x 69 columns]

In [51]:
df_flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 121884 entries, 0 to 121883
Data columns (total 69 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   YEAR                   121884 non-null  int64  
 1   QUARTER                121884 non-null  int64  
 2   MONTH                  121884 non-null  int64  
 3   DAY_OF_MONTH           121884 non-null  int64  
 4   DAY_OF_WEEK            121884 non-null  int64  
 5   FL_DATE                121884 non-null  object 
 6   OP_UNIQUE_CARRIER      121884 non-null  object 
 7   OP_CARRIER_AIRLINE_ID  121884 non-null  int64  
 8   TAIL_NUM               120394 non-null  object 
 9   ORIGIN_AIRPORT_ID      121884 non-null  int64  
 10  ORIGIN_AIRPORT_SEQ_ID  58000 non-null   float64
 11  ORIGIN_CITY_MARKET_ID  121884 non-null  int64  
 12  ORIGIN                 121884 non-null  object 
 13  ORIGIN_CITY_NAME       58000 non-null   object 
 14  ORIGIN_STATE_ABR       121884 non-nu

In [55]:
count_cancellation = df_flights["CANCELLATION_CODE"].isna().sum()
count_cancellation

119085

The cancellation code column has a lot of na values, so we will have to drop it

In [58]:
df_flights_clean = df_flights.drop('CANCELLATION_CODE', axis = 1)

In [60]:
df_flights_clean.shape
df_flights_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 121884 entries, 0 to 121883
Data columns (total 68 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   YEAR                   121884 non-null  int64  
 1   QUARTER                121884 non-null  int64  
 2   MONTH                  121884 non-null  int64  
 3   DAY_OF_MONTH           121884 non-null  int64  
 4   DAY_OF_WEEK            121884 non-null  int64  
 5   FL_DATE                121884 non-null  object 
 6   OP_UNIQUE_CARRIER      121884 non-null  object 
 7   OP_CARRIER_AIRLINE_ID  121884 non-null  int64  
 8   TAIL_NUM               120394 non-null  object 
 9   ORIGIN_AIRPORT_ID      121884 non-null  int64  
 10  ORIGIN_AIRPORT_SEQ_ID  58000 non-null   float64
 11  ORIGIN_CITY_MARKET_ID  121884 non-null  int64  
 12  ORIGIN                 121884 non-null  object 
 13  ORIGIN_CITY_NAME       58000 non-null   object 
 14  ORIGIN_STATE_ABR       121884 non-nu

In [61]:
df_flights_without_na = df_flights_clean.dropna()

In [62]:
df_flights_without_na.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9449 entries, 15 to 57964
Data columns (total 68 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   YEAR                   9449 non-null   int64  
 1   QUARTER                9449 non-null   int64  
 2   MONTH                  9449 non-null   int64  
 3   DAY_OF_MONTH           9449 non-null   int64  
 4   DAY_OF_WEEK            9449 non-null   int64  
 5   FL_DATE                9449 non-null   object 
 6   OP_UNIQUE_CARRIER      9449 non-null   object 
 7   OP_CARRIER_AIRLINE_ID  9449 non-null   int64  
 8   TAIL_NUM               9449 non-null   object 
 9   ORIGIN_AIRPORT_ID      9449 non-null   int64  
 10  ORIGIN_AIRPORT_SEQ_ID  9449 non-null   float64
 11  ORIGIN_CITY_MARKET_ID  9449 non-null   int64  
 12  ORIGIN                 9449 non-null   object 
 13  ORIGIN_CITY_NAME       9449 non-null   object 
 14  ORIGIN_STATE_ABR       9449 non-null   object 
 15  OR

Droping na values results in a significant data loss, do we drop all na values? How should we handle it.